go-controller/pkg/kubevirt/router.go

```diff
@@ -0,0 +1,325 @@
+		return ownsItAndIsOrphanOrWrongZone(item.ExternalIDs, vms)
+	}); err != nil {
+		return fmt.Errorf("failed deleting stale vm policies: %v", err)
+	}
+	if err := libovsdbops.DeleteDHCPOptionsWithPredicate(nbClient, func(item *nbdb.DHCPOptions) bool {
+		return ownsItAndIsOrphanOrWrongZone(item.ExternalIDs, vms)
+	}); err != nil {
+		return fmt.Errorf("failed deleting stale vm policies: %v", err)
+	}
+	return nil
+}
+
+func DeleteRoutingForMigratedPodWithZone(nbClient libovsdbclient.Client, pod *corev1.Pod, zone string) error {
+	vm := ExtractVMNameFromPod(pod)
+	predicate := func(itemExternalIDs map[string]string) bool {
+		containsZone := true
+		if zone != "" {
+			containsZone = itemExternalIDs[OvnZoneExternalIDKey] == zone
+		}
+		return containsZone && externalIDsContainsVM(itemExternalIDs, vm)
+	}
+	routePredicate := func(item *nbdb.LogicalRouterStaticRoute) bool {
+		return predicate(item.ExternalIDs)
+	}
+	if err := libovsdbops.DeleteLogicalRouterStaticRoutesWithPredicate(nbClient, types.OVNClusterRouter, routePredicate); err != nil {
+		return fmt.Errorf("failed deleting pod routing when deleting the LR static routes: %v", err)
+	}
+	policyPredicate := func(item *nbdb.LogicalRouterPolicy) bool {
+		return predicate(item.ExternalIDs)
+	}
+	if err := libovsdbops.DeleteLogicalRouterPoliciesWithPredicate(nbClient, types.OVNClusterRouter, policyPredicate); err != nil {
+		return fmt.Errorf("failed deleting pod routing when deleting the LR policies: %v", err)
+	}
+	return nil
+}
+
+func DeleteRoutingForMigratedPod(nbClient libovsdbclient.Client, pod *corev1.Pod) error {
+	return DeleteRoutingForMigratedPodWithZone(nbClient, pod, "")
+}
+
+// EnsureLocalZonePodAddressesToNodeRoute will add static routes and policies to ovn_cluster_route logical router
+// to ensure VM traffic work as expected after live migration if the pod is running at the local/global zone.
+//
+// NOTE: IC with multiple nodes per zone is not supported
+//
+// Following is the list of NB logical resources created depending if it's interconnected or not:
+//
+// IC (on node per zone):
+//   - static route with cluster wide CIDR as src-ip prefix and nexthop GR, it has less
+//     priority than route to use overlay in case of pod to pod communication
+//
+// NO IC:
+//   - low priority policy with src VM ip and reroute GR, since it has low priority
+//     it will not override the policy to enroute pod to pod traffic using overlay
+//
+// Both:
+//   - static route with VM ip as dst-ip prefix and output port the LRP pointing to the VM's node switch
+func EnsureLocalZonePodAddressesToNodeRoute(watchFactory *factory.WatchFactory, nbClient libovsdbclient.Client, pod *corev1.Pod, nadName string) error {
+	vmReady, err := virtualMachineReady(watchFactory, pod)
+	if err != nil {
+		return err
+	}
+	if !vmReady {
+		return nil
+	}
+	podAnnotation, err := util.UnmarshalPodAnnotation(pod.Annotations, nadName)
```

Comment:
This will fail at least once right? Because we don't see yet the pod updated with the annotation.

Perhaps this is ok because `virtualMachineReady` will always return false at first.
Or perhaps `EnsureLocalZonePodAddressesToNodeRoute` could be called from addLogicalPort

Not extremely concerned about this right now

Reply:
Another option is to use the LSP addresses instead of relying on the pod annotation

Reply:
> This will fail at least once right? Because we don't see yet the pod updated with the annotation.
> 
> Perhaps this is ok because `virtualMachineReady` will always return false at first. Or perhaps `EnsureLocalZonePodAddressesToNodeRoute` could be called from addLogicalPort
> 
> Not extremely concerned about this right now

As you mention the annotation is added before migrated VM is ready, so this will be called again and the annotation will be there

> Another option is to use the LSP addresses instead of relying on the pod annotation

As I understand correctly at ovn-k in general is better to calculate stuff from etcd than from ovsdb, also reading from LSP will not add much value since vm readiness mechanism will repeat the function.